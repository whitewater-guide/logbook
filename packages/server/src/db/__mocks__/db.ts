import {
  DatabasePoolConnectionType,
  DatabasePoolType,
  createPool,
} from 'slonik';

import { Client } from 'pg';

const cfg = {
  host: process.env.POSTGRES_HOST,
  user: process.env.POSTGRES_USER,
  database: `logbook_test_${process.env.JEST_WORKER_ID}`,
  password: process.env.POSTGRES_PASSWORD,
};

let _pool: DatabasePoolType | undefined;

export const db: () => DatabasePoolConnectionType = () => {
  if (!_pool) {
    throw new Error('no test pool found');
  }
  return _pool;
};

export const setupDB = async () => {
  const workerID = process.env.JEST_WORKER_ID;
  if (!workerID) {
    throw new Error('JEST_WORKER_ID not found');
  }
  const client = new Client({
    host: process.env.POSTGRES_HOST,
    user: process.env.POSTGRES_USER,
    database: process.env.POSTGRES_DB,
    password: process.env.POSTGRES_PASSWORD,
  });
  try {
    await client.connect();
    await client.query(`DROP DATABASE IF EXISTS logbook_test_${workerID}`);
    await client.query(
      `CREATE DATABASE logbook_test_${workerID} TEMPLATE logbook_test_template`,
    );
  } finally {
    await client.end();
  }

  _pool = createPool(
    `postgresql://${cfg.user}:${cfg.password}@${cfg.host}/${cfg.database}`,
  );
};

export const teardownDB = async () => {
  await _pool?.end();
};
