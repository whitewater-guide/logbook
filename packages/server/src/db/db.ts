import { createPool, DatabasePoolConnectionType } from 'slonik';

const cfg = {
  host: process.env.POSTGRES_HOST,
  user: process.env.POSTGRES_USER,
  database: process.env.POSTGRES_DB,
  password: process.env.POSTGRES_PASSWORD,
};

const pool = createPool(
  `postgresql://${cfg.user}:${cfg.password}@${cfg.host}/${cfg.database}`,
);

export const db: () => DatabasePoolConnectionType = () => pool;
export const setupDB = () => Promise.resolve();
export const teardownDB = async () => {
  await pool.end();
};
