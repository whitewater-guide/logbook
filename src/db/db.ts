import {
  DatabasePoolConnectionType,
  createPool,
  createTypeParserPreset,
} from 'slonik';

const cfg = {
  host: process.env.POSTGRES_HOST,
  user: process.env.POSTGRES_USER,
  database: process.env.POSTGRES_DB,
  password: process.env.POSTGRES_PASSWORD,
};

const typeParsers = createTypeParserPreset().filter(
  ({ name }) => !['name', 'timestamp', 'timestamptz'].includes(name),
);

const pool = createPool(
  `postgresql://${cfg.user}:${cfg.password}@${cfg.host}/${cfg.database}`,
  {
    typeParsers,
  },
);

export const db: () => DatabasePoolConnectionType = () => pool;
export const setupDB = () => Promise.resolve();
export const teardownDB = async () => {
  await pool.end();
};
