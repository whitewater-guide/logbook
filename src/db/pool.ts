import { createPool, createTypeParserPreset } from 'slonik';

const cfg = {
  host: process.env.POSTGRES_HOST,
  user: process.env.POSTGRES_USER,
  database:
    process.env.NODE_ENV === 'test'
      ? `logbook_test_${process.env.JEST_WORKER_ID}`
      : process.env.POSTGRES_DB,
  password: process.env.POSTGRES_PASSWORD,
};

const typeParsers = createTypeParserPreset().filter(
  ({ name }) => !['name', 'timestamp', 'timestamptz'].includes(name),
);

export const db = createPool(
  `postgresql://${cfg.user}:${cfg.password}@${cfg.host}/${cfg.database}`,
  {
    typeParsers,
  },
);
