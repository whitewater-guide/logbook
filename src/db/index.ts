import { db } from './pool';

export default db;
