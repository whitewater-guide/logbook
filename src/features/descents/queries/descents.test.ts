import {
  DESCENT_1,
  DESCENT_10,
  DESCENT_2,
  DESCENT_3,
  DESCENT_4,
  DESCENT_5,
  DESCENT_7,
  SECTION_1,
  SECTION_1_UPSTREAM,
  USER_1,
  USER_2,
} from '~/test/fixtures';
import {
  ListDescentsQuery,
  ListDescentsQueryVariables,
} from './descents.test.generated';
import { setupDB, teardownDB } from '~/db';

import DescentFragments from '../descents.fragments';
import { DescentsFilter } from '~/__generated__/graphql';
import { gql } from 'apollo-server';
import { runQuery } from '~/test/apollo-helpers';

beforeEach(setupDB);
afterEach(teardownDB);

const query = gql`
  query listDescents($filter: DescentsFilter, $page: Page) {
    descents(filter: $filter, page: $page) {
      edges {
        node {
          ...descentAll
        }
        cursor
      }
      pageInfo {
        endCursor
        hasMore
      }
    }
  }
  ${DescentFragments.All}
`;

const getIds = (result: ListDescentsQuery): string[] =>
  result.descents?.edges.map(({ node }) => node.id) || [];

it('should match snapshot', async () => {
  const result = await runQuery<ListDescentsQuery, ListDescentsQueryVariables>(
    query,
    {},
    USER_1,
  );
  expect(result).toMatchSnapshot();
});

it.each([
  ['anon should not see private sections', undefined, false],
  ['non-owner should not see private sections', USER_2, false],
  ['owner should see private sections', USER_1, true],
])('%s', async (_, uid, visible) => {
  const result = await runQuery<ListDescentsQuery, ListDescentsQueryVariables>(
    query,
    {},
    uid,
  );
  const actual = getIds(result.data).includes(DESCENT_2);
  expect(actual).toBe(visible);
});

type FilterTestCase = [string, DescentsFilter, string[]];
// use snapshot as a baseline
it.each<FilterTestCase>([
  ['difficulty', { difficulty: [2.5, 3] }, [DESCENT_7, DESCENT_3, DESCENT_2]],
  [
    'open date range',
    { startDate: new Date('2020-01-03T00:00:00.000Z') },
    [DESCENT_7, DESCENT_5, DESCENT_3],
  ],
  [
    'closed date range',
    {
      startDate: new Date('2020-01-02T00:00:00.000Z'),
      endDate: new Date('2020-01-05T00:00:00.000Z'),
    },
    [DESCENT_5, DESCENT_3, DESCENT_2],
  ],
  [
    'user',
    {
      userID: USER_1,
    },
    [DESCENT_3, DESCENT_2, DESCENT_1],
  ],
  [
    'section id', // including derived sections
    {
      sectionID: SECTION_1,
    },
    [DESCENT_10, DESCENT_4, DESCENT_1],
  ],
  [
    'upstream section id',
    {
      upstreamSectionID: SECTION_1_UPSTREAM,
    },
    [DESCENT_10, DESCENT_4, DESCENT_1],
  ],
  [
    'section (full) name',
    {
      sectionName: 'oR',
    },
    [
      DESCENT_7,
      DESCENT_5,
      DESCENT_3,
      DESCENT_2,
      DESCENT_10,
      DESCENT_4,
      DESCENT_1,
    ],
  ],
])(
  `should filter and paginate by %s`,
  async (_, filter, [id1, id2, ...restIds]) => {
    let result = await runQuery<ListDescentsQuery, ListDescentsQueryVariables>(
      query,
      { filter, page: { limit: 1 } },
      USER_1,
    );
    expect(result.errors).toBeUndefined();
    expect(getIds(result.data)).toEqual([id1]);
    expect(result.data.descents?.pageInfo.hasMore).toBe(true);
    result = await runQuery<ListDescentsQuery, ListDescentsQueryVariables>(
      query,
      {
        filter,
        page: { limit: 1, after: result.data.descents?.pageInfo.endCursor },
      },
      USER_1,
    );
    expect(result.errors).toBeUndefined();
    expect(getIds(result.data)).toEqual([id2]);
    expect(result.data.descents?.pageInfo.hasMore).toBe(true);
    result = await runQuery<ListDescentsQuery, ListDescentsQueryVariables>(
      query,
      {
        filter,
        page: { limit: 99, after: result.data.descents?.pageInfo.endCursor },
      },
      USER_1,
    );
    expect(result.errors).toBeUndefined();
    expect(getIds(result.data)).toEqual(restIds);
    expect(result.data.descents?.pageInfo.hasMore).toBe(false);
  },
);
