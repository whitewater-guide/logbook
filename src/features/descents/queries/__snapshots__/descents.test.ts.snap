// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
Object {
  "data": Object {
    "descents": Object {
      "edges": Array [
        Object {
          "cursor": "djE6NzoxNTc4MzU1MjAwMDAw",
          "node": Object {
            "comment": "descent 7 comment",
            "createdAt": 2020-01-07T00:00:00.000Z,
            "duration": null,
            "id": "ac323bc5-47e7-43cf-b3dd-e029cf96fcf4",
            "level": Object {
              "unit": "cm",
              "value": 211,
            },
            "public": true,
            "section": Object {
              "createdAt": 2020-01-07T00:00:00.000Z,
              "difficulty": 2.5,
              "id": "2186e9bf-19f8-4a26-a9a1-80de877c4084",
              "putIn": null,
              "region": "Georgia",
              "river": "Bzhuzha",
              "section": "Lower",
              "takeOut": null,
              "updatedAt": 2020-01-07T00:00:00.000Z,
              "upstreamData": null,
              "upstreamId": null,
            },
            "startedAt": 2020-01-07T00:00:00.000Z,
            "updatedAt": 2020-01-07T00:00:00.000Z,
            "upstreamData": null,
            "userId": "8e295235-d6aa-4fbf-bf64-1f9c9decba5c",
          },
        },
        Object {
          "cursor": "djE6NToxNTc4MTgyNDAwMDAw",
          "node": Object {
            "comment": "descent 5 comment",
            "createdAt": 2020-01-05T00:00:00.000Z,
            "duration": 1000,
            "id": "3a12cd33-2ccd-4fcd-9da5-7ce14006aa15",
            "level": Object {
              "unit": null,
              "value": 4,
            },
            "public": true,
            "section": Object {
              "createdAt": 2020-01-05T00:00:00.000Z,
              "difficulty": 3.5,
              "id": "eae06053-a542-46d7-9708-6d08f1284d7f",
              "putIn": null,
              "region": "Mexico",
              "river": "Alseseca",
              "section": "Roadside",
              "takeOut": null,
              "updatedAt": 2020-01-05T00:00:00.000Z,
              "upstreamData": null,
              "upstreamId": null,
            },
            "startedAt": 2020-01-05T00:00:00.000Z,
            "updatedAt": 2020-01-05T00:00:00.000Z,
            "upstreamData": null,
            "userId": "60d37885-4d41-4ce2-9ae6-e413eaa28ed3",
          },
        },
        Object {
          "cursor": "djE6MzoxNTc4MDA5NjAwMDAw",
          "node": Object {
            "comment": "descent 3 comment",
            "createdAt": 2020-01-03T00:00:00.000Z,
            "duration": null,
            "id": "7deddd48-9f7f-4a76-9ff9-811f694f52a4",
            "level": Object {
              "unit": null,
              "value": null,
            },
            "public": true,
            "section": Object {
              "createdAt": 2020-01-03T00:00:00.000Z,
              "difficulty": 2.5,
              "id": "7565d9ee-0fc0-4b49-be34-a62d59223067",
              "putIn": null,
              "region": "Norway",
              "river": "Sjoa",
              "section": "Asengjuvet",
              "takeOut": null,
              "updatedAt": 2020-01-03T00:00:00.000Z,
              "upstreamData": null,
              "upstreamId": null,
            },
            "startedAt": 2020-01-03T00:00:00.000Z,
            "updatedAt": 2020-01-03T00:00:00.000Z,
            "upstreamData": null,
            "userId": "ad808d80-7d1d-4bee-8276-0b9204e46b2f",
          },
        },
        Object {
          "cursor": "djE6MjoxNTc3OTIzMjAwMDAw",
          "node": Object {
            "comment": "descent 2 comment",
            "createdAt": 2020-01-02T00:00:00.000Z,
            "duration": 3000,
            "id": "3fb2b1da-6649-429e-8b0e-0cbbda055da7",
            "level": Object {
              "unit": "m",
              "value": 0.9,
            },
            "public": false,
            "section": Object {
              "createdAt": 2020-01-02T00:00:00.000Z,
              "difficulty": 3,
              "id": "e7d03e01-4927-4ae1-82bf-a9e195e59db6",
              "putIn": null,
              "region": "Norway",
              "river": "Sjoa",
              "section": "Playrun",
              "takeOut": null,
              "updatedAt": 2020-01-02T00:00:00.000Z,
              "upstreamData": null,
              "upstreamId": null,
            },
            "startedAt": 2020-01-02T00:00:00.000Z,
            "updatedAt": 2020-01-02T00:00:00.000Z,
            "upstreamData": null,
            "userId": "ad808d80-7d1d-4bee-8276-0b9204e46b2f",
          },
        },
        Object {
          "cursor": "djE6MTA6MTU3NzgzNjgwMDAwMA==",
          "node": Object {
            "comment": "descent 10, parent descent 1 comment",
            "createdAt": 2020-01-10T00:00:00.000Z,
            "duration": 3600,
            "id": "7f712c8a-45a9-4fd5-966a-c8b4d07f36df",
            "level": Object {
              "unit": "m3/s",
              "value": 99.9,
            },
            "public": true,
            "section": Object {
              "createdAt": 2020-01-09T00:00:00.000Z,
              "difficulty": 3.5,
              "id": "720f5a61-a272-4552-a731-1d1d9848f9b7",
              "putIn": null,
              "region": "Norway",
              "river": "Sjoa",
              "section": "Amot",
              "takeOut": null,
              "updatedAt": 2020-01-09T00:00:00.000Z,
              "upstreamData": Object {
                "upstream": "upstream section 1 data",
              },
              "upstreamId": "d749dbc1-166f-41d1-91ac-e4658d30d90f",
            },
            "startedAt": 2020-01-01T00:00:00.000Z,
            "updatedAt": 2020-01-10T00:00:00.000Z,
            "upstreamData": Object {
              "upstream": "descent 1 upstream data",
            },
            "userId": "8e295235-d6aa-4fbf-bf64-1f9c9decba5c",
          },
        },
        Object {
          "cursor": "djE6NDoxNTc3ODM2ODAwMDAw",
          "node": Object {
            "comment": "descent 4, parent descent 1 comment",
            "createdAt": 2020-01-04T00:00:00.000Z,
            "duration": 3600,
            "id": "ffd6ba41-d7e5-4fbf-8347-2481d583674e",
            "level": Object {
              "unit": "m3/s",
              "value": 99.9,
            },
            "public": true,
            "section": Object {
              "createdAt": 2020-01-04T00:00:00.000Z,
              "difficulty": 4,
              "id": "f9035355-0499-4d1c-9ccd-c9909dbad49a",
              "putIn": null,
              "region": "Norway",
              "river": "Sjoa",
              "section": "Amot",
              "takeOut": null,
              "updatedAt": 2020-01-04T00:00:00.000Z,
              "upstreamData": Object {
                "upstream": "upstream section 1 data",
              },
              "upstreamId": "d749dbc1-166f-41d1-91ac-e4658d30d90f",
            },
            "startedAt": 2020-01-01T00:00:00.000Z,
            "updatedAt": 2020-01-04T00:00:00.000Z,
            "upstreamData": Object {
              "upstream": "descent 1 upstream data",
            },
            "userId": "60d37885-4d41-4ce2-9ae6-e413eaa28ed3",
          },
        },
        Object {
          "cursor": "djE6MToxNTc3ODM2ODAwMDAw",
          "node": Object {
            "comment": "descent 1 comment",
            "createdAt": 2020-01-01T00:00:00.000Z,
            "duration": 3600,
            "id": "97998f67-74f2-43db-a4e4-198c2ca540a7",
            "level": Object {
              "unit": "m3/s",
              "value": 99.9,
            },
            "public": true,
            "section": Object {
              "createdAt": 2020-01-01T00:00:00.000Z,
              "difficulty": 3.5,
              "id": "45e70501-5d20-449f-b854-52b06b138831",
              "putIn": Object {
                "lat": 49,
                "lng": 0,
              },
              "region": "Norway",
              "river": "Sjoa",
              "section": "Amot",
              "takeOut": Object {
                "lat": 49,
                "lng": 1,
              },
              "updatedAt": 2020-01-01T00:00:00.000Z,
              "upstreamData": Object {
                "upstream": "upstream section 1 data",
              },
              "upstreamId": "d749dbc1-166f-41d1-91ac-e4658d30d90f",
            },
            "startedAt": 2020-01-01T00:00:00.000Z,
            "updatedAt": 2020-01-01T00:00:00.000Z,
            "upstreamData": Object {
              "upstream": "descent 1 upstream data",
            },
            "userId": "ad808d80-7d1d-4bee-8276-0b9204e46b2f",
          },
        },
      ],
      "pageInfo": Object {
        "endCursor": "djE6MToxNTc3ODM2ODAwMDAw",
        "hasMore": false,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
